src/rpc/methods/eth.rs

```diff
@@ -1338,13 +1330,84 @@ impl RpcMethod<2> for EthGetCode {
         let get_bytecode_return: GetBytecodeReturn =
             fvm_ipld_encoding::from_slice(msg_rct.return_data().as_slice())?;
         if let Some(cid) = get_bytecode_return.0 {
-            Ok(Bytes(ctx.store().get_required(&cid)?))
+            Ok(EthBytes(ctx.store().get_required(&cid)?))
         } else {
             Ok(Default::default())
         }
     }
 }
 
+pub enum EthGetStorageAt {}
+impl RpcMethod<3> for EthGetStorageAt {
+    const NAME: &'static str = "Filecoin.EthGetStorageAt";
+    const PARAM_NAMES: [&'static str; 3] = ["eth_address", "position", "block_number_or_hash"];
+    const API_VERSION: ApiVersion = ApiVersion::V1;
+    const PERMISSION: Permission = Permission::Read;
+
+    type Params = (EthAddress, EthBytes, BlockNumberOrHash);
+    type Ok = EthBytes;
+
+    async fn handle(
+        ctx: Ctx<impl Blockstore + Send + Sync + 'static>,
+        (eth_address, position, block_number_or_hash): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let make_empty_result = || EthBytes(vec![0; 32]);
+
+        let ts = tipset_by_block_number_or_hash(&ctx.chain_store, block_number_or_hash)?;
+        let to_address = FilecoinAddress::try_from(&eth_address)?;
+        let Some(actor) = ctx
+            .state_manager
+            .get_actor(&to_address, *ts.parent_state())?
+        else {
+            return Ok(make_empty_result());
+        };
+
+        if !fil_actor_interface::is_evm_actor(&actor.code) {
+            return Ok(make_empty_result());
+        }
+
+        let params = RawBytes::new(GetStorageAtParams::new(position.0)?.serialize_params()?);
+        let message = Message {
+            from: FilecoinAddress::SYSTEM_ACTOR,
+            to: to_address,
+            method_num: METHOD_GET_STORAGE_AT,
+            gas_limit: BLOCK_GAS_LIMIT,
+            params,
+            ..Default::default()
+        };
+        let mut api_invoc_result = None;
+        for ts in ts.chain_arc(ctx.store()) {
+            match ctx.state_manager.call(&message, Some(ts)) {
+                Ok(res) => {
+                    api_invoc_result = Some(res);
+                    break;
+                }
+                Err(e) => tracing::warn!(%e),
+            }
+        }
+        let Some(api_invoc_result) = api_invoc_result else {
```

Comment:
Looking at this more I'm thinking: why do we need `Option` for `api_invoc_result`? It would be much easier to use `Result`. Then there is no need for `let Some..`

Something like:

```
  for ... {
    match {
       Ok(res) -> return Ok(res)
    } 
  }
  return Err(err)

  and then just do `api_invoc_result?`
  
```

Reply:
I don't think the `return` clause works this way. https://github.com/ChainSafe/forest/pull/4340#discussion_r1615529393

Reply:
Fixed.